### PR TITLE
Use result.Result in more places.

### PR DESCRIPTION
**Pull request by CyrusNajmabadi**

### CODE REVIEW

File: pkg/engine/plan.go [hunk 1]
```diff
@@ -242,13 +243,17 @@ func (planResult *planResult) Close() error {
...
 	// Walk the plan's steps and and pretty-print them out.
 	actions := newPlanActions(planResult.Options)
-	if err := planResult.Walk(ctx, actions, true); err != nil {
-		return nil, errors.New("an error occurred while advancing the preview")
+	if res := planResult.Walk(ctx, actions, true); res != nil {
+		if res.IsBail() {
+			return nil, res
+		}
+
+		return nil, result.Error("an error occurred while advancing the preview")
```
Comment (CyrusNajmabadi) [inline pkg/engine/plan.go]: unfortunate.  it seems like we've lost the original err information (this was a problem before).

File: sdk/nodejs/tests/runtime/langhost/run.spec.ts [hunk 2]
```diff
@@ -727,10 +727,15 @@ describe("rpc", () => {
                 return { urn: makeUrn(t, name), id: undefined, props: undefined };
             },
         },
+        "run_error": {
+            program: path.join(base, "040.run_error"),
+            expectResourceCount: 0,
+            expectError: "Program exited with non-zero exit code: 1",
```
Comment (CyrusNajmabadi) [inline sdk/nodejs/tests/runtime/langhost/run.spec.ts]: this error will change in a followup PR.

File: pkg/engine/plan.go (hunk 1 above)
Comment (swgillespie) [reply pkg/engine/plan.go]: I think that's OK - we should have already printed it, right?

Comment (swgillespie) [reply pkg/engine/plan.go]: Oh, nevermind - in this case, we won't have printed it. Could you instead return `res` verbatim without creating the new error?

File: pkg/resource/deploy/plan_executor.go [hunk 3]
```diff
@@ -189,13 +193,22 @@ func (pe *planExecutor) Execute(callerCtx context.Context, opts Options, preview
...
+	var err error
+	if res != nil {
+		err = res.Error()
+	}
+
 	// Figure out if execution failed and why. Step generation and execution errors trump cancellation.
 	if err != nil || pe.stepExec.Errored() {
 		err = execError("failed", preview)
 	} else if canceled {
 		err = execError("canceled", preview)
```
Comment (swgillespie) [inline pkg/resource/deploy/plan_executor.go]: I'm thinking this should be a bail as well; cancellation doesn't sound to me like something that should be a non-bail Result.

File: pkg/resource/deploy/source_eval.go [hunk 4]
```diff
@@ -151,14 +154,19 @@ func (iter *evalSourceIterator) Next() (SourceEvent, error) {
 		contract.Assert(read != nil)
 		logging.V(5).Infoln("EvalSourceIterator produced a read")
 		return read, nil
-	case err := <-iter.finChan:
+	case res := <-iter.finChan:
 		// If we are finished, we can safely exit.  The contract with the language provider is that this implies
 		// that the language runtime has exited and so calling Close on the plugin is fine.
+		// fmt.Printf("evalSourceIterator.next.  got error and returned it: %v\n", err)
```
Comment (swgillespie) [inline pkg/resource/deploy/source_eval.go]: nit:

```suggestion
```

**Comment by CyrusNajmabadi** [comment]
Talked to @swgillespie .  Will take on some of hte feedback later when we can be sure we don't make any scenarios worse.  specificaly, we don't wnat to end up printing out the same error multiple times.

**Pull request merged**